### ISSUE TITLE
add ability to cancel long-running query

### ISSUE THREAD

**Issue by mispp** [mispp/goat#26]
need a way to cancel long-running query.
`m_sqlQuery.finish();` and `m_queryFuture.cancel();` are not doing anything

option could maybe be to kill a cloned (!) connection

**Comment by asw-dev** [comment]
I ran into this as well in my own testing with threads. The only way to "cancel" a call to QSqlQuery::exec is to run it on another thread and call QThread::terminate on it. Unfortunately this risks leaking memory / database resources so it's not really an acceptable solution.

http://www.qtcentre.org/threads/23905-Abort-QSqlQuery-exec()?p=115888#post115888

For that matter, threading sql in qt seams a little troublesome as according to the docs anything short of having the full open/query/close on the same thread is unsupported and might break in the future.

http://doc.qt.io/archives/qt-4.8/threads-modules.html

It would be possible to fake a cancel by disconnecting the QThread::finish so the results don't go anywhere, but that still has a thread & database session running. For a real solution you will probably need to use a different library.

**Comment by mispp** [comment]
Yeah, this is going to be an issue for sure, so moving to libpq is not so outdated afterall.

I asked on stackoverflow how to deal with this problem and the answer is [here](https://stackoverflow.com/questions/51801513/how-to-cancel-long-running-qsqlquery)

Long story short - using QtSql, we need workarounds (which might be ok for the time being).

> For that matter, threading sql in qt seams a little troublesome as according to the docs anything short of having the full open/query/close on the same thread is unsupported and might break in the future.

This is now a bit more worrying for me since i wanted to do the following: 

Right now, QSqlQuery is a member of QueryTab. It gets worked on by a new thread which is spawned upon clicking run which executes it. Once the thread is finished, this object is still in the QueryTab and it is used by another thread - a thread that would populate QStandardItemModel or QAbstractTableModel, whichever gets to be used (thinking the latter makes more sense).
Since looping over results is needed to make this work, i was going to do data retrieval in again a new thread and load a chunk of 1000 rows into the model. Once the QTableView is scrolled to the very end, another load for next 1000 rows would be triggered.

Can't think of a better way to do it since i cant understand how fetchMode would be used in this...

**Comment by mispp** [comment]
Anyhow, query cancellation for postgres is commited. Something similar can be done for sure with mysql.

**Comment by asw-dev** [comment]
The `pg_backend_pid()` & `pg_cancel_backend()` approach should be able to work for any major db except sqlite. You might consider removing the call to finish() and let the exec-ing thread handled it when it is unblocked as that is probably the source of the "connection pointer is NULL" message.

The approach I was working can do same-thread queries. The idea is you can call ConnectionManager::query(connectionId) which instead of returning a QSqlQuery returns a QueryThread that will do all the open setup when the thread starts, The thread exposes the QAbstractTableModel that will be written to (manually w/ forward-only) and has finished & failed signals. It is started/canceled by the tab. If left at that it would still be holding/rendering the full set of data, but you could in the future have the data written to a memory mapped temp file to support paging on large datasets (which is kinda like what the os does with memory already until you run out of swap space). This uses more memory/disk, but closes the connection as fast as possible.

There are a lot of conflicts with my changes, and it is currently not in a working state so I will probably end up dropping them rather than attempt a fix & rebase.

**Comment by mispp** [comment]
just tried dropping QSqlQuery::finish(), but still the error is here (if you meant this one).

for how it currently works... i have no idea if this is actually good practice. i did it in a way i found logical, but i do know that some refactoring will be needed. any huge objections on how things are done now? 

some things like cloning a connection should definitely be moved to conneciton manager thought (will be done at some point).

**Comment by asw-dev** [comment]
I have no objections.

**Comment by mispp** [comment]
you mentioned that you did already do some work on this... do you have maybe subclassed QAbstractTableModel that you made work? I'm trying to set it up now but if you did it already, maybe i can just reuse it

**Comment by asw-dev** [comment]
Hadn't got that far, but I whipped up an example after reviewing the docs. Turns out QStandardItemModel can represent a table so there is no need to subclass.

Here is the rough code (use what you want, change the rest).

```
QStandardItemModel* model = &m_model;
model->clear();
for (int row = 0; query.next(); ++row)
{
    QSqlRecord record = query.record();
    if (row == 0)
    {
        model->setColumnCount(record.count());
        for (int col = 0; col < record.count(); ++col)
        {
            model->setHeaderData(col, Qt::Horizontal, record.fieldName(col));
        }
    }

    model->setRowCount(model->rowCount() + 1);
    for (int col = 0; col < record.count(); ++col)
    {
        model->setData(model->index(row, col), record.value(col));
    }
}
ui->resultsGrid->setModel(model);
```
I didn't look into how to handle header data when there are 0 rows.

**Comment by mispp** [comment]
Thanks! Will try it this way.

**Comment by asw-dev** [comment]
Just added QMYSQL to my dev branch (no PR due to conflicts with the differences in thread impl.). The queries you are looking for are `SELECT CONNECTION_ID();` and `KILL QUERY :id;`

https://github.com/asw-dev/goat/commit/6d4464ce03e9167a7709d3d577248adc23550769

**Comment by mispp** [comment]
So this can be generalized, great. I'll add it this week.

For the threading, i'll also have to take a look at your implementation.  It might make sense to rework what i did since your QueryStates are a bit more verbose. QFuture supports several of them though...

**Comment by mispp** [comment]
this is now implemented for 3 drivers POSTGRES, MYSQL and ODBC.
